### DISCUSSION TITLE
Visual representation of workflows

### DISCUSSION THREAD

**Discussion by johnkm516** [infiniticio/infinitic#162] [Ideas] [answered]
Are there plans for a visual representation of workflows? A workflow would be code-first as it is currently, but if infinitic could generate a visual state machine for each execution it would greatly help in understanding each workflow, what state the workflow was in at each step, which would help debug as well.

I'm looking for an orchestration framework to use for my company's ERP that we're building from the ground up. I'm currently considering AWS Step Functions, Temporal, and Infinitic, but I am quite set on Infinitic because : 

1) It uses Apache Pulsar which is open source and practically guarantees resilience and no data loss, with battle-tested geosync features that work (compared to Temporal which I believe still has issues with its XDC replication which has made Netflix create their own Temporal SDK extensions). 

2) It's a code-first workflow framework which means all the features that can be used in code without any limitations can be used in workflows such as dynamic, generic workflows

3) It's not vendor-locked and it's open source and free unlike AWS Step Functions. 

The one advantage that Step Functions has and draws a lot of people is its visual representation of workflows. While Step Functions' design choice is to orchestrate flows entirely in a visual manner, I want to develop workflows in code but have a visual representation of workflows that I could use to debug and look at without having to refer back to source code. 

Just thinking about it I think it would be a pretty colossal task, but I'm wondering if Infinitic has any plans for this kind of feature in the far future.

**Comment by geomagilles** [comment, marked as answer]
Sorry for the delay in responding - I missed your question - not sure why.

It is relatively straightforward:
1) to use/define a DSL to define workflows operated by Infinitic (I can put you in touch with a company doing it), 
2) to display the execution details of a specific workflow instance

But providing a visual representation of a workflow from its code would be rather complex. It would be very similar to taking an arbitrary code and visually representing it. A specialized AST parser could do that - but it is a complex subject, and the output may not be as good as expected.

Luckily, workflow code is usually simple enough to be self-explaining. If you need a visual correspondence, I would suggest working on the 1) above - e.g., Infinitic could be used to operate workflows defined in the same DSL as Step Functions (which is a subset of what Infinitic is capable of)
